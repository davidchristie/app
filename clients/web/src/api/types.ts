export interface Api {
  useAuthorize(): (providerId: string) => void;
  useSession: UseQuery<Session>;
}

export interface Session {
  user: User | null;
}

export interface QueryResult<Data> {
  data?: Data;
  error?: Error;
  loading: boolean;
}

export interface UseQuery<Data> {
  (): QueryResult<Data>;
}

export interface User {
  id: string;
  name: string;
  email: string;
}
