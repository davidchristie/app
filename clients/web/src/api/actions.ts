export function authorize(providerId: string): void {
  window.location.href = `/api/v1/auth/${providerId}/authorize`;
}

export function signOut(): void {
  window.location.href = "/api/v1/auth/signout";
}
