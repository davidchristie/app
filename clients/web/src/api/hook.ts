import { useContext } from "react";
import { ApiContext } from "./context";
import { Api } from "./types";

export function useApi(): Api {
  const api = useContext(ApiContext);
  if (!api) {
    throw new Error("API context not found");
  }
  return api;
}
