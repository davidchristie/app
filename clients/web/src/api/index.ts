export * from "./hook";
export * from "./mock";
export * from "./provider";
export * from "./types";
