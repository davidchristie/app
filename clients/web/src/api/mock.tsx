import { ReactNode } from "react";
import { ApiContext } from "./context";
import { Api, Session, UseQuery } from "./types";

export interface MockApiData {
  authorize?: () => void;
  session?: Session | Error;
}

export interface MockApiProviderProps {
  children?: ReactNode;
  data?: MockApiData;
}

function createQueryHook<Data>(
  value: Data | Error | undefined
): UseQuery<Data> {
  return () => {
    if (!value) {
      return {
        loading: true,
      };
    } else if (value instanceof Error) {
      return {
        error: value,
        loading: false,
      };
    }
    return {
      data: value,
      loading: false,
    };
  };
}

export function MockApiProvider({
  children,
  data = {},
}: MockApiProviderProps): JSX.Element {
  const api: Api = {
    useAuthorize: () => data.authorize ?? jest.fn(),
    useSession: createQueryHook(data.session),
  };
  return <ApiContext.Provider value={api}>{children}</ApiContext.Provider>;
}
