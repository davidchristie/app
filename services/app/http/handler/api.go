package handler

import (
	"net/http"

	"github.com/davidchristie/app/services/app/auth"
	"github.com/davidchristie/app/services/app/http/middleware"
	"github.com/go-chi/chi/v5"
)

func apiHandler(auth auth.Auth) http.Handler {
	r := chi.NewRouter()
	r.Use(middleware.Session(auth))
	r.Mount("/v1", restV1Handler(auth))
	return r
}

func restV1Handler(auth auth.Auth) http.Handler {
	r := chi.NewRouter()
	r.Get("/auth/session", Session())
	r.Get("/auth/{providerID}/authorize", Authorize(auth))
	r.Get("/auth/{providerID}/callback", OAuthCallback(auth))
	r.Get("/auth/signout", SignOut())
	return r
}
