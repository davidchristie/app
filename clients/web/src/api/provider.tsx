import { ReactNode } from "react";
import useSWR from "swr";
import { authorize } from "./actions";
import { ApiContext } from "./context";
import { Api, UseQuery } from "./types";

export interface ApiProviderProps {
  children?: ReactNode;
}

function createQueryHook<Data>(url: string): UseQuery<Data> {
  return () => {
    const { data, error } = useSWR(url);
    const loading = !error && !data;
    return { data, error, loading };
  };
}

export function ApiProvider({ children }: ApiProviderProps): JSX.Element {
  const api: Api = {
    useAuthorize: () => authorize,
    useSession: createQueryHook("/api/v1/auth/session"),
  };
  return <ApiContext.Provider value={api}>{children}</ApiContext.Provider>;
}
