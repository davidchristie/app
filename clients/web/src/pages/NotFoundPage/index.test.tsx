import { MockApiProvider } from "../../api";
import { render, screen } from "../../testing";
import { signedInSession } from "../../testing/data";
import { NotFoundPage } from ".";

describe("NotFoundPage", () => {
  it("renders correctly", () => {
    const { container } = render(
      <MockApiProvider data={{ session: signedInSession }}>
        <NotFoundPage />
      </MockApiProvider>
    );
    const element = screen.getByTestId("NotFoundPage");
    expect(element).toBe(container.firstChild);
    expect(element).toMatchSnapshot();
  });
});
