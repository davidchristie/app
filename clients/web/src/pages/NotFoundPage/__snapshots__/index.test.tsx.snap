// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFoundPage renders correctly 1`] = `
<div
  class="MuiBox-root MuiBox-root-1"
  data-testid="NotFoundPage"
>
  <div
    class="MuiBox-root MuiBox-root-2"
    data-testid="Page"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorDefault MuiPaper-elevation4"
      data-testid="AppBar"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <div
          class="MuiBox-root MuiBox-root-5"
        >
          <a
            class="MuiTypography-root makeStyles-title-3 MuiTypography-h6"
            href="/"
          >
            App
          </a>
          <div
            class="MuiBox-root MuiBox-root-6"
          >
            <div
              class="MuiBox-root MuiBox-root-7"
              data-testid="UserMenu"
            >
              <button
                aria-controls="user-menu"
                aria-haspopup="true"
                class="MuiButtonBase-root MuiIconButton-root"
                data-testid="UserMenu__iconButton"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <div
                    class="MuiAvatar-root MuiAvatar-circular"
                  >
                    <img
                      class="MuiAvatar-img"
                      src="https://via.placeholder.com/150"
                    />
                  </div>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </header>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg"
      data-testid="MainContent"
    >
      <div
        class="MuiBox-root MuiBox-root-8"
      >
        <h2
          class="MuiTypography-root MuiTypography-h2"
          data-testid="PageHeading"
        >
          Not Found
        </h2>
      </div>
    </div>
  </div>
</div>
`;
